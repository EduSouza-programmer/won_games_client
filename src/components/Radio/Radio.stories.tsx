import { Story, Meta } from "@storybook/react/types-6-0";
import Radio, { RadioProps } from ".";

export default {
  title: "Radio",
  component: Radio,
  argTypes: {
    labelColor: {
      options: ["white", "black"],
      control: { type: "radio" },
    },
    labelFor: {
      control: false,
    },
    onCheck: { action: "checked" },
    value: {
      control: { type: "text" },
    },
  },
  parameters: {
    viewport: {
      defaultViewport: "mobile1",
    },
    backgrounds: {
      default: "won-dark",
    },
    layout: "fullscreen",
  },
} as Meta<RadioProps>;

export const Default: Story<RadioProps> = (args) => (
  <div
    style={{
      width: "100%",
      minHeight: "100vh",
      display: "flex",
      flexDirection: "column",
      justifyContent: "center",
      alignItems: "center",
    }}
  >
    <div
      style={{
        width: "50%",
        height: "10rem",
        display: "flex",
        flexDirection: "column",
        justifyContent: "space-between",
      }}
    >
      <Radio
        name="radio"
        label="Maior preço"
        labelFor="maior preço"
        value="primeiro"
        defaultChecked
      />
      <Radio
        name="radio"
        label="Menor preço"
        labelFor="menor preço"
        value="segundo"
      />
      <Radio
        name="radio"
        label="TRY IT"
        labelFor="try it"
        value="try it"
        {...args}
      />
    </div>
  </div>
);
