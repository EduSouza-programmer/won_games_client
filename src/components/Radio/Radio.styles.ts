import styled, { css } from "styled-components";

import { RadioProps } from ".";

export const Wrapper = styled.div`
  display: flex;
  align-items: center;
`;

export const Radio = styled.input`
  ${({ theme }) => css`
    display: flex;
    justify-content: center;
    align-items: center;
    position: relative;

    appearance: none;
    width: 1.8rem;
    height: 1.8rem;
    border: 0.2rem solid ${theme.palette.primary.main};
    border-radius: 50%;
    cursor: pointer;
    outline: none;

    &::before {
      content: "";
      width: 0.8rem;
      height: 0.8rem;
      background: ${theme.palette.primary.main};
      border-radius: 50%;
      position: absolute;
      opacity: 0;
      transform: scale(0);
      transition: all 0.2s ease-in-out;
    }

    &:checked {
      &::before {
        opacity: 1;
        transform: scale(1);
      }
    }

    &:focus {
      box-shadow: 0 0 0.5rem ${theme.palette.primary.main};
    }
  `}
`;

export const Label = styled.label<Pick<RadioProps, "labelColor">>`
  ${({ theme, labelColor }) => css`
    color: ${theme.palette[labelColor!].main};
    margin-left: ${theme.spacings.xxsmall};
    line-height: 1;
  `}
`;
