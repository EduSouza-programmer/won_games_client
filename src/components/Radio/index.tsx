import { InputHTMLAttributes, useRef, useCallback } from "react";
import * as S from "./Radio.styles";

type RadioValue = string | ReadonlyArray<string> | number;

export type RadioProps = {
  label?: string;
  labelFor?: string;
  labelColor?: "white" | "black";
  onCheck?: (radioValue?: RadioValue) => void;
  value?: RadioValue;
} & InputHTMLAttributes<HTMLInputElement>;

function Radio({
  label,
  labelFor,
  labelColor = "white",
  onCheck,
  value,
  ...props
}: RadioProps) {
  const inputRef = useRef<HTMLInputElement>(null);

  const onChange = useCallback(() => {
    if (onCheck) {
      onCheck(inputRef.current?.value);
    }
  }, [onCheck]);
  return (
    <S.Wrapper>
      <S.Radio
        onChange={onChange}
        ref={inputRef}
        type="radio"
        id={labelFor}
        value={value}
        {...props}
      />
      {!!label && (
        <S.Label labelColor={labelColor} htmlFor={labelFor}>
          {label}
        </S.Label>
      )}
    </S.Wrapper>
  );
}

export default Radio;
