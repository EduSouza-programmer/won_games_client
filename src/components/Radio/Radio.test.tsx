import { screen, waitFor } from "@testing-library/react";
import { renderWithTheme } from "@/utils/tests/helpers";
import userEvent from "@testing-library/user-event";

import Radio from ".";

describe("<Radio />", () => {
  it("should render with label", () => {
    const { container } = renderWithTheme(
      <Radio label="Radio label" labelFor="radio" />
    );

    expect(screen.getByRole("radio")).toBeInTheDocument();
    expect(screen.getByLabelText(/radio label/i)).toBeInTheDocument();
    expect(screen.getByText(/radio label/i)).toHaveAttribute("for", "radio");

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render without label", () => {
    renderWithTheme(<Radio />);
    expect(screen.getByRole("radio").parentElement?.childElementCount).toBe(1);
  });

  it("should render label color white by default", () => {
    renderWithTheme(<Radio label="my label" labelFor="radio" />);

    expect(screen.getByText(/my label/i)).toHaveStyle({
      color: "#FAFAFA",
    });
  });
  it("should render label color black when passed", () => {
    renderWithTheme(
      <Radio label="my label" labelFor="radio" labelColor="black" />
    );

    expect(screen.getByText(/my label/i)).toHaveStyle({
      color: "#030517",
    });
  });
  it("should dispatch onCheck when click event happen", async () => {
    const onCheck = jest.fn();
    renderWithTheme(
      <Radio
        label="my label"
        labelFor="radio"
        onCheck={onCheck}
        value="my value"
      />
    );

    expect(onCheck).not.toHaveBeenCalled();

    userEvent.click(screen.getByLabelText(/my label/i));
    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1);
    });
    expect(onCheck).toHaveBeenCalledWith("my value");
  });

  it("should be accessible with tab", async () => {
    renderWithTheme(<Radio label="my label" labelFor="radio" />);

    expect(document.body).toHaveFocus();

    userEvent.tab();
    expect(screen.getByLabelText(/my label/i)).toHaveFocus();
    userEvent.tab();
    expect(screen.getByLabelText(/my label/)).not.toHaveFocus();
  });
});
