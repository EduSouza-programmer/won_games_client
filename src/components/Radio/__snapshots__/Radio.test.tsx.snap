// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Radio /> should render with label 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  width: 1.8rem;
  height: 1.8rem;
  border: 0.2rem solid #F231A5;
  border-radius: 50%;
  cursor: pointer;
  outline: none;
}

.c1::before {
  content: "";
  width: 0.8rem;
  height: 0.8rem;
  background: #F231A5;
  border-radius: 50%;
  position: absolute;
  opacity: 0;
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
}

.c1:checked::before {
  opacity: 1;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c1:focus {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c2 {
  color: #FAFAFA;
  margin-left: 0.8rem;
  line-height: 1;
}

<div
  class="c0"
>
  <input
    class="c1"
    id="radio"
    type="radio"
    value=""
  />
  <label
    class="c2"
    for="radio"
  >
    Radio label
  </label>
</div>
`;
